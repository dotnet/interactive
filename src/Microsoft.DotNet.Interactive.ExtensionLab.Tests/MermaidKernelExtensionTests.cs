// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;

using Assent;

using FluentAssertions;
using FluentAssertions.Execution;
using HtmlAgilityPack;
using Microsoft.DotNet.Interactive.Commands;
using Microsoft.DotNet.Interactive.CSharp;
using Microsoft.DotNet.Interactive.Events;
using Microsoft.DotNet.Interactive.Formatting;
using Microsoft.DotNet.Interactive.Tests.Utility;

using Xunit;

namespace Microsoft.DotNet.Interactive.ExtensionLab.Tests
{
    public class MermaidKernelExtensionTests : IDisposable
    {
        private readonly Configuration _configuration;

        public MermaidKernelExtensionTests()
        {
            _configuration = new Configuration()
                .UsingExtension("txt")
                .SetInteractive(Debugger.IsAttached);
        }

        [Fact]
        public async Task adds_mermaid_kernel()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            kernel.ChildKernels
                .Should()
                .ContainSingle(k => k is MermaidKernel);
        }

        [Fact]
        public async Task registers_html_formatter_for_MermaidMarkdown()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            var markdown = @"graph TD
    A[Client] --> B[Load Balancer]
    B --> C[Server1]
    B --> D[Server2]";

            var formatted = new MermaidMarkdown(markdown).ToDisplayString(HtmlFormatter.MimeType);
            var doc = new HtmlDocument();
            doc.LoadHtml(formatted.FixedGuid());
            var scriptNode = doc.DocumentNode.SelectSingleNode("//div/script");
            var renderTarget = doc.DocumentNode.SelectSingleNode("//div[@id='00000000000000000000000000000000']");
            using var _ = new AssertionScope();

            scriptNode.Should().NotBeNull();
            scriptNode.InnerText.Should()
                .Contain(markdown);
            scriptNode.InnerText.Should()
                .Contain("configureRequireFromExtension('Mermaid','1.0.0')(['Mermaid/mermaidapi'], (mermaid) => {");

            renderTarget.Should().NotBeNull();
        }

        [Fact]
        public async Task registers_html_formatter_for_explorer()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            var explorer = typeof(List<string>).ExploreWithUmlClassDiagram();
            var formatted = explorer.ToDisplayString(HtmlFormatter.MimeType);
            var markdown = explorer.ToMarkdown().ToString();

            var doc = new HtmlDocument();
            doc.LoadHtml(formatted.FixedGuid());
            var scriptNode = doc.DocumentNode.SelectSingleNode("//div/script");
            var renderTarget = doc.DocumentNode.SelectSingleNode("//div[@id='00000000000000000000000000000000']");
            
            using var _ = new AssertionScope();

            scriptNode.Should().NotBeNull();
            scriptNode.InnerText.Should()
                .Contain(markdown);
            scriptNode.InnerText.Should()
                .Contain("configureRequireFromExtension('Mermaid','1.0.0')(['Mermaid/mermaidapi'], (mermaid) => {");

            renderTarget.Should().NotBeNull();
        }


        [Fact]
        public async Task mermaid_kernel_handles_SubmitCode()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            KernelCommandResult result = await kernel.SubmitCodeAsync(@"
#!mermaid
    graph TD
    A[Client] --> B[Load Balancer]
    B --> C[Server1]
    B --> D[Server2]
");

            var events = result.KernelEvents.ToSubscribedList();

            var formattedData = events
                .OfType<DisplayedValueProduced>()
                .Single()
                .FormattedValues
                .Single(fm => fm.MimeType == HtmlFormatter.MimeType)
                .Value;

            this.Assent(formattedData.FixedGuid(), _configuration);
        }

        [Fact]
        public async Task can_use_extension_methods_from_the_kernel_extension()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            await kernel.SendAsync(new SubmitCode($@"#r ""{typeof(MermaidKernelExtension).Assembly.Location}""", "csharp"));

            var result = await kernel.SendAsync( new SubmitCode(@"
using System;

typeof(List<string>).ExploreWithUmlClassDiagram().Display();
", "csharp"));

            var events = result.KernelEvents.ToSubscribedList();

            var formattedData = events
                .OfType<DisplayedValueProduced>()
                .Single()
                .FormattedValues
                .Single(fm => fm.MimeType == HtmlFormatter.MimeType)
                .Value;

            this.Assent(formattedData.FixedGuid(), _configuration);
        }

        [Fact]
        public void can_generate_class_diagram_from_type()
        {
            var diagram = typeof(JsonElement).ToClassDiagram();

            this.Assent(diagram.ToString());
        }

        [Fact]
        public void can_generate_class_diagram_from_generic_type()
        {
            var diagram = typeof(List<Dictionary<string,object>>).ToClassDiagram();

            this.Assent(diagram.ToString());
        }


        [Fact]
        public void can_generate_class_diagram_to_a_specified_depth()
        {
            var diagram = typeof(List<Dictionary<string, object>>).ToClassDiagram(new ClassDiagramConfiguration(1));

            diagram.ToString().Should()
                .NotContain(
                    "ICollection~Dictionary<String, Object>~ --|> IEnumerable~Dictionary<String, Object>~ : Inheritance");
        }

        [Fact]
        public void can_explore_a_type_with_UmlClassDiagram()
        {
            var diagram = typeof(List<Dictionary<string, object>>).ExploreWithUmlClassDiagram().ToMarkdown();

            this.Assent(diagram.ToString());
        }

        [Fact]
        public void can_configure_UmlClassDiagramExplorer()
        {
            var diagram = typeof(List<Dictionary<string, object>>).ExploreWithUmlClassDiagram()
                .WithGraphDepth(1).ToMarkdown();
            
            diagram.ToString().Should()
                .NotContain(
                    "ICollection~Dictionary<String, Object>~ --|> IEnumerable~Dictionary<String, Object>~ : Inheritance");
        }

        public void Dispose()
        {
            Formatter.ResetToDefault();
        }
    }
}