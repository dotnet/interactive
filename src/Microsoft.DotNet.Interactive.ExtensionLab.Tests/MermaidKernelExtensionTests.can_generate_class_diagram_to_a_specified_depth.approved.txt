classDiagram
CSharpKernel --|> DotNetKernel : Inheritance
CSharpKernel --|> IDisposable : Inheritance
CSharpKernel --|> IExtensibleKernel : Inheritance
CSharpKernel --|> ISupportNuget : Inheritance
CSharpKernel --|> IKernelCommandHandler~RequestCompletions~ : Inheritance
CSharpKernel --|> IKernelCommandHandler~RequestDiagnostics~ : Inheritance
CSharpKernel --|> IKernelCommandHandler~RequestHoverText~ : Inheritance
CSharpKernel --|> IKernelCommandHandler~RequestSignatureHelp~ : Inheritance
CSharpKernel --|> IKernelCommandHandler~SubmitCode~ : Inheritance
DotNetKernel --|> Kernel : Inheritance
DotNetKernel --|> IDisposable : Inheritance


class CSharpKernel
CSharpKernel : +get_ScriptState() ScriptState
CSharpKernel : -set_ScriptState(ScriptState value) Void
CSharpKernel : +IsCompleteSubmissionAsync(String code) Task~Boolean~
CSharpKernel : +GetVariableNames() IReadOnlyCollection~String~
CSharpKernel : +TryGetVariable(String name, T& value) Boolean
CSharpKernel : +SetVariableAsync(String name, Object value, Type declaredType) Task
CSharpKernel : +HandleAsync(RequestHoverText command, KernelInvocationContext context) Task
CSharpKernel : +HandleAsync(RequestSignatureHelp command, KernelInvocationContext context) Task
CSharpKernel : +HandleAsync(SubmitCode submitCode, KernelInvocationContext context) Task
CSharpKernel : -RunAsync(String code, CancellationToken cancellationToken, Func~Exception, Boolean~ catchException) Task
CSharpKernel : +HandleAsync(RequestCompletions command, KernelInvocationContext context) Task
CSharpKernel : -GetCompletionList(String code, Int32 cursorPosition) Task~IEnumerable<CompletionItem>~
CSharpKernel : GetDiagnosticsProduced(KernelCommand command, ImmutableArray~Diagnostic~ diagnostics) DiagnosticsProduced
CSharpKernel : +HandleAsync(RequestDiagnostics command, KernelInvocationContext context) Task
CSharpKernel : +LoadExtensionsFromDirectoryAsync(DirectoryInfo directory, KernelInvocationContext context) Task
CSharpKernel : +get_PackageRestoreContext() PackageRestoreContext
CSharpKernel : -get_HasReturnValue() Boolean
CSharpKernel : -Microsoft.DotNet.Interactive.ISupportNuget.AddRestoreSource(String source) Void
CSharpKernel : -Microsoft.DotNet.Interactive.ISupportNuget.GetOrAddPackageReference(String packageName, String packageVersion) PackageReference
CSharpKernel : -Microsoft.DotNet.Interactive.ISupportNuget.RegisterResolvedPackageReferences(IReadOnlyList~ResolvedPackageReference~ resolvedReferences) Void
CSharpKernel : -Microsoft.DotNet.Interactive.ISupportNuget.RestoreAsync() Task~PackageRestoreResult~
CSharpKernel : +get_RequestedPackageReferences() IEnumerable~PackageReference~
CSharpKernel : +get_ResolvedPackageReferences() IEnumerable~ResolvedPackageReference~
CSharpKernel : +get_RestoreSources() IEnumerable~String~
CSharpKernel : -<.ctor>b__8_0() PackageRestoreContext
CSharpKernel : -<.ctor>b__8_1() Void

class DotNetKernel
DotNetKernel : +TryGetVariable(String name, T& value) Boolean*
DotNetKernel : +SetVariableAsync(String name, Object value, Type declaredType) Task*
DotNetKernel : +GetVariableNames() IReadOnlyCollection~String~*

class IDisposable
<<interface>> IDisposable
IDisposable : +Dispose() Void*

class IExtensibleKernel
<<interface>> IExtensibleKernel
IExtensibleKernel : +LoadExtensionsFromDirectoryAsync(DirectoryInfo directory, KernelInvocationContext invocationContext) Task*

class ISupportNuget
<<interface>> ISupportNuget
ISupportNuget : +RestoreAsync() Task~PackageRestoreResult~*
ISupportNuget : +get_RestoreSources() IEnumerable~String~*
ISupportNuget : +get_RequestedPackageReferences() IEnumerable~PackageReference~*
ISupportNuget : +get_ResolvedPackageReferences() IEnumerable~ResolvedPackageReference~*
ISupportNuget : +GetOrAddPackageReference(String packageName, String packageVersion) PackageReference*
ISupportNuget : +AddRestoreSource(String source) Void*
ISupportNuget : +RegisterResolvedPackageReferences(IReadOnlyList~ResolvedPackageReference~ packageReferences) Void*

class IKernelCommandHandler~RequestCompletions~
<<interface>> IKernelCommandHandler~RequestCompletions~
IKernelCommandHandler~RequestCompletions~ : +HandleAsync(RequestCompletions command, KernelInvocationContext context) Task*

class IKernelCommandHandler~RequestDiagnostics~
<<interface>> IKernelCommandHandler~RequestDiagnostics~
IKernelCommandHandler~RequestDiagnostics~ : +HandleAsync(RequestDiagnostics command, KernelInvocationContext context) Task*

class IKernelCommandHandler~RequestHoverText~
<<interface>> IKernelCommandHandler~RequestHoverText~
IKernelCommandHandler~RequestHoverText~ : +HandleAsync(RequestHoverText command, KernelInvocationContext context) Task*

class IKernelCommandHandler~RequestSignatureHelp~
<<interface>> IKernelCommandHandler~RequestSignatureHelp~
IKernelCommandHandler~RequestSignatureHelp~ : +HandleAsync(RequestSignatureHelp command, KernelInvocationContext context) Task*

class IKernelCommandHandler~SubmitCode~
<<interface>> IKernelCommandHandler~SubmitCode~
IKernelCommandHandler~SubmitCode~ : +HandleAsync(SubmitCode command, KernelInvocationContext context) Task*

class Kernel
Kernel : get_Pipeline() KernelCommandPipeline
Kernel : +get_ParentKernel() CompositeKernel
Kernel : set_ParentKernel(CompositeKernel value) Void
Kernel : +get_SubmissionParser() SubmissionParser
Kernel : +AddMiddleware(KernelCommandPipelineMiddleware middleware, String caller) Void
Kernel : +DeferCommand(KernelCommand command) Void
Kernel : -TryPreprocessCommands(KernelCommand originalCommand, KernelInvocationContext context, IReadOnlyList commands) Boolean
Kernel : -TryPreprocessLanguageServiceCommand(LanguageServiceCommand command, KernelInvocationContext context, IReadOnlyList commands) Boolean
Kernel : +get_FrontendEnvironment() FrontendEnvironment
Kernel : +set_FrontendEnvironment(FrontendEnvironment value) Void
Kernel : get_IsDisposed() Boolean
Kernel : +get_KernelEvents() IObservable~KernelEvent~
Kernel : +get_Name() String
Kernel : +set_Name(String value) Void
Kernel : get_Uri() KernelUri
Kernel : +get_Directives() IReadOnlyCollection~ICommand~
Kernel : +AddDirective(Command command) Void
Kernel : +RegisterCommandHandler(Func~TCommand, KernelInvocationContext, Task~ handler) Void
Kernel : +RegisterCommandType() Void
Kernel : HandleAsync(KernelCommand command, KernelInvocationContext context) Task
Kernel : +SendAsync(KernelCommand command, CancellationToken cancellationToken) Task~KernelCommandResult~
Kernel : InvokePipelineAndCommandHandler(KernelCommand command) Task~KernelCommandResult~
Kernel : get_Scheduler() KernelScheduler~KernelCommand, KernelCommandResult~
Kernel : SetScheduler(KernelScheduler~KernelCommand, KernelCommandResult~ scheduler) Void
Kernel : GetHandlingKernelUri(KernelCommand command) KernelUri
Kernel : PublishEvent(KernelEvent kernelEvent) Void
Kernel : +RegisterForDisposal(Action dispose) Void
Kernel : +RegisterForDisposal(IDisposable disposable) Void
Kernel : -HandleRequestCompletionsAsync(RequestCompletions command, KernelInvocationContext context) Task
Kernel : -GetDirectiveCompletionItems(DirectiveNode directiveNode, Int32 requestPosition) IEnumerable~CompletionItem~
Kernel : GetDirectiveParsersForCompletion(DirectiveNode directiveNode, Int32 requestPosition) IEnumerable~Parser~
Kernel : TrySetHandler(KernelCommand command, KernelInvocationContext context) Void
Kernel : -TrySetDynamicHandler(KernelCommand command) Void
Kernel : -SetHandler(IKernelCommandHandler~T~ handler, T command) Void$
Kernel : SetHandlingKernel(KernelCommand command, KernelInvocationContext context) Void
Kernel : +Dispose() Void
Kernel : CreateChooseKernelDirective() ChooseKernelDirective
Kernel : get_ChooseKernelDirective() ChooseKernelDirective
Kernel : +get_Current() Kernel$
Kernel : +get_Root() Kernel$
Kernel : +display(Object value, String mimeType) DisplayedValue$
Kernel : +HTML(String content) IHtmlContent$
Kernel : +CSS(String content) Void$
Kernel : +Javascript(String scriptContent) Void$
Kernel : -<.ctor>b__7_0() Void
Kernel : -<SetScheduler>g__GetDeferredOperations|45_0(KernelCommand command, String scope) IEnumerable~KernelCommand~

